// reflex/lib/application.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.start = exports.beginner = undefined;

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }();

var _task = require("./task");

var _signal = require("./signal");

var _effects = require("./effects");

var _dom = require("./dom");

var first = function first(xs) {
  return xs[0];
};
var second = function second(xs) {
  return xs[1];
};

var beginner = exports.beginner = function beginner(configuration) {
  return { flags: void 0,
    init: function init(_) {
      return [configuration.model, _effects.Effects.none];
    },
    update: function update(model, action) {
      return [configuration.update(model, action), _effects.Effects.none];
    },
    view: configuration.view
  };
};

var start = exports.start = function start(configuration) {
  var init = configuration.init;
  var view = configuration.view;
  var update = configuration.update;
  var flags = configuration.flags;

  var _mailbox = (0, _signal.mailbox)({});

  var address = _mailbox.address;
  var signal = _mailbox.signal;

  var step = function step(_ref, action) {
    var _ref2 = _slicedToArray(_ref, 2);

    var model = _ref2[0];
    var _ = _ref2[1];
    return update(model, action);
  };

  var display = function display(model) {
    return (0, _dom.root)(view, model, address);
  };

  var steps = (0, _signal.reductions)(step, init(flags), signal);

  var model = (0, _signal.map)(first, steps);
  var application = { address: address,
    model: model,
    task: (0, _signal.map)(second, steps),
    view: (0, _signal.map)(display, model)
  };

  return application;
};
